File: gallery/interview/leetcode/1143. 最长公共子序列.py
# https://leetcode-cn.com/problems/longest-common-subsequence/

# https://leetcode-cn.com/problems/maximum-length-of-repeated-subarray/



# 执行结果：
# 通过
# 显示详情
# 执行用时：
# 440 ms
# , 在所有 Python3 提交中击败了
# 77.17%
# 的用户
# 内存消耗：
# 34.4 MB
# , 在所有 Python3 提交中击败了
# 5.03%
# 的用户

    


class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        # find the length of the strings 
        s1 = text1
        s2 = text2
        
        m = len(s1) 
        n = len(s2) 

        dp = [ [None] * (n + 1) for _ in range(m + 1) ] # m+1;n+1
        prev =  [ [ (None, None) for __ in range(n + 1)] for _ in range(m + 1) ]
        

        for i in range(m + 1): # i [0,m] dp[0][0] 表示不考虑s1 和s2任何字符; dp[2][2]表示考虑数学意义上的s1前2个字符和s2前2个字符
            for j in range(n + 1): 
                if i == 0 or j == 0 : 
                    dp[i][j] = 0
                                    
                elif s1[i-1] == s2[j-1]: 
                    dp[i][j] = dp[i-1][j-1]+1
                    prev[i][j] = [i-1, j-1]
                else:
                    if dp[i-1][j] > dp[i][j-1]: # 暂时不考虑多解的情况
                        dp[i][j] = dp[i-1][j]
                    else:
                        dp[i][j] = dp[i][j-1]
        
        return dp[m][n]

File: gallery/interview/leetcode/1143. 最长公共子序列.cpp
/*
https://leetcode-cn.com/problems/longest-common-subsequence/
执行结果：
通过

执行用时：
16 ms, 在所有 C++ 提交中击败了89.72%的用户
内存消耗：
10.5 MB, 在所有 C++ 提交中击败了85.29%的用户
通过测试用例：
44 / 44

class Solution {
public:
    int dp[1005][1005];
    int longestCommonSubsequence(string text1, string text2) {
        int m = text1.size();
        int n = text2.size();
        int max_len = 0;
        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; j++){
                if(text1[i-1] == text2[j-1]){
                    dp[i][j] = max(dp[i][j], dp[i-1][j-1] + 1);
                    max_len = max(max_len, dp[i][j]);
                }
                else{
                    dp[i][j] = max(dp[i][j], max(dp[i-1][j], dp[i][j-1]));
                }
            }
        }
        return max_len;
    }
};

*/